ecs_client_test.go

```diff
@@ -39,49 +60,205 @@ func TestECSClientRegisterAndDeregisterTaskDefinition(t *testing.T) {
+			assert.Zero(t, out)
+		},
+		"RunTaskFailsWithValidButNonexistentInput": func(ctx context.Context, t *testing.T, c *BasicECSClient) {
+			out, err := c.RunTask(ctx, &ecs.RunTaskInput{
+				TaskDefinition: aws.String(utility.RandomString()),
+			})
+			assert.Error(t, err)
+			assert.Zero(t, out)
+		},
+		"DescribeTasksFailsWithNoClusterAndValidButNonexistentInput": func(ctx context.Context, t *testing.T, c *BasicECSClient) {
+			out, err := c.DescribeTasks(ctx, &ecs.DescribeTasksInput{
+				Tasks: []*string{aws.String(utility.RandomString())},
+			})
+			assert.Error(t, err)
+			assert.Zero(t, out)
+		},
+		"DescribeTasksReturnsFailureWithClusterAndValidButNonexistentInput": func(ctx context.Context, t *testing.T, c *BasicECSClient) {
+			out, err := c.DescribeTasks(ctx, &ecs.DescribeTasksInput{
+				Cluster: aws.String(os.Getenv("AWS_ECS_CLUSTER")),
+				Tasks:   []*string{aws.String("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa")},
+			})
+
+			assert.NoError(t, err)
+			assert.NotZero(t, out)
+			assert.NotZero(t, out.Failures)
+			assert.Empty(t, out.Tasks)
+
+		},
+		"StopTaskFailsWithValidButNonexistentInput": func(ctx context.Context, t *testing.T, c *BasicECSClient) {
+			out, err := c.StopTask(ctx, &ecs.StopTaskInput{
+				Task: aws.String(utility.RandomString()),
+			})
+			assert.Error(t, err)
+			assert.Zero(t, out)
+		},
+		"RegisterSucceedsWithDuplicateTaskDefinition": func(ctx context.Context, t *testing.T, c *BasicECSClient) {
+			outDuplicate, err := c.RegisterTaskDefinition(ctx, &ecs.RegisterTaskDefinitionInput{
+				ContainerDefinitions: []*ecs.ContainerDefinition{
+					{
+						Command: []*string{aws.String("echo"), aws.String("foo")},
+						Image:   aws.String("ubuntu"),
+						Name:    aws.String("print_foo"),
+					},
+				},
+				Cpu:    aws.String("128"),
+				Memory: aws.String("4"),
+				Family: aws.String(os.Getenv("AWS_ECS_TASK_DEFINITION_PREFIX") + "bar"),
+			})
+
+			require.NoError(t, err)
+			require.NotZero(t, outDuplicate)
+			require.NotZero(t, outDuplicate.TaskDefinition)
+
+			defer cleanupTaskDefinition(ctx, t, c, outDuplicate)
+
+			assert.True(t, *outDuplicate.TaskDefinition.Revision > *registerOut.TaskDefinition.Revision)
+
+		},
+		"RunAndStopTaskSucceedsWithRegisteredTaskDefinition": func(ctx context.Context, t *testing.T, c *BasicECSClient) {
+
+			require.NotZero(t, registerOut.TaskDefinition.Status)
+			assert.Equal(t, "ACTIVE", *registerOut.TaskDefinition.Status)
+
+			runOut, err := c.RunTask(ctx, &ecs.RunTaskInput{
+				Cluster:        aws.String(os.Getenv("AWS_ECS_CLUSTER")),
+				TaskDefinition: registerOut.TaskDefinition.TaskDefinitionArn,
+			})
+
+			require.NoError(t, err)
+			require.NotZero(t, runOut)
 
-		defer func() {
-			if out != nil && out.TaskDefinition != nil && out.TaskDefinition.TaskDefinitionArn != nil {
-				out, err := c.DeregisterTaskDefinition(ctx, &ecs.DeregisterTaskDefinitionInput{
-					TaskDefinition: out.TaskDefinition.TaskDefinitionArn,
+			defer func() {
```

Comment:
Isn't this identical to `cleanupTask`?

Reply:
Yup, just refactored the code to avoid using `defer`